src/daft-local-execution/src/streaming_pipeline.rs

```diff
@@ -0,0 +1,185 @@
+            data = op.execute(&data)?;
+        }
+        let _ = send_to_sink.send(Ok(data)).await;
+        println!("Intermediate operators finished");
+        Ok(())
+    }
+
+    async fn launch_sink(
+        mut receivers: Vec<Receiver>,
+        mut sink: Option<Box<dyn Sink>>,
+        send_to_next_source: Sender,
+    ) -> DaftResult<()> {
+        println!("Launching sink");
+        let in_order = match sink {
+            Some(ref sink) => sink.in_order(),
+            None => false,
+        };
+
+        if in_order {
+            for receiver in receivers.iter_mut() {
+                while let Some(val) = receiver.recv().await {
+                    if let Some(sink) = sink.borrow_mut() {
+                        let sink_result = sink.sink(&val?)?;
+                        match sink_result {
+                            SinkResultType::Finished => {
+                                break;
+                            }
+                            SinkResultType::NeedMoreInput => {}
+                        }
+                    } else {
+                        let _ = send_to_next_source.send(val).await;
+                    }
+                }
+            }
+        } else {
+            let mut unordered_receivers = FuturesUnordered::new();
+            for receiver in receivers.iter_mut() {
+                unordered_receivers.push(receiver.recv());
+            }
+            while let Some(val) = unordered_receivers.next().await {
+                if let Some(val) = val {
+                    if let Some(sink) = sink.borrow_mut() {
+                        let sink_result = sink.sink(&val?)?;
+                        match sink_result {
+                            SinkResultType::Finished => {
+                                break;
+                            }
+                            SinkResultType::NeedMoreInput => {}
+                        }
+                    } else {
+                        let _ = send_to_next_source.send(val).await;
+                    }
+                }
+            }
+        }
+
+        if let Some(sink) = sink.borrow_mut() {
+            println!("Finalizing sink");
+            let final_values = sink.finalize()?;
+            for value in final_values {
+                println!("Sending finalized value back to pipeline");
+                let _ = send_to_next_source.send(Ok(value)).await;
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn run(
+        send_to_next_source: Sender,
+        source: Box<dyn Source>,
+        intermediate_operators: Vec<Box<dyn IntermediateOperator>>,
+        sink: Option<Box<dyn Sink>>,
+    ) -> DaftResult<()> {
+        println!("Running pipeline");
+
+        let source_stream = source.get_data();
+        pin_mut!(source_stream);
+
+        let mut handles = FuturesUnordered::new();
+        let mut receivers = vec![];
+        while let Some(morsel) = source_stream.next().await {
```

Comment:
There should be a fixed number of parallel pipelines that are pushed data to in round robin fashion